test/external/external_llama_eval.py

```diff
@@ -1,102 +1,77 @@
-    assert isinstance(model_size, str)
-    assert isinstance(batch_size, int)
-    assert isinstance(checkpoint_path, str)
-    assert isinstance(tokenizer_path, str)
-
-    self.llama = LLaMa.build(checkpoint_path, tokenizer_path, model_gen, model_size, quantize)
-
-  @classmethod
-  def create_from_arg_string(cls, arg_string, additional_config=None):
-    kwargs = {el.split("=")[0]: el.split("=")[1] for el in arg_string.split(",")}
-    return cls(**kwargs, **additional_config)
-
-  @property
-  def eot_token_id(self):
-    # we use EOT because end of *text* is more accurate for what we're doing than end of *sentence*
-    return self.llama.tokenizer.eos_id()
-
-  @property
-  def max_length(self):
-    return 1024
-
+    self.max_length = max_length
+    self.tokenizer = Tokenizer(str((checkpoint_path if checkpoint_path.is_dir() else checkpoint_path.parent) / "tokenizer.model"))
+    self.model = build_transformer(checkpoint_path, model_size=model_size, quantize=quantize, max_context=self.max_length)
   @property
-  def max_gen_toks(self):
-    return 256
-
-  @property
-  def batch_size(self):
-    return 1
-
-  @property
-  def device(self):
-    return self._device
-
-  def tok_encode(self, string: str):
-    return [self.llama.tokenizer.bos_id()] + self.llama.tokenizer.encode(string)
-
-  def tok_decode(self, tokens):
-    return self.llama.tokenizer.decode(tokens)
-
-  def _model_call(self, inps):
-    return torch.Tensor(self.llama.model(Tensor(inps.numpy()), 0).numpy())
-
-  def greedy_until(self, requests):
+  def tokenizer_name(self) -> str: pass
+  def chat_template(self, chat_template: bool | str = False) -> str: pass
+  def apply_chat_template(self, chat_history: list[dict[str, str]], add_generation_prompt: bool = True) -> str:
+    ret = ""
+    for message in chat_history:
+      ret += f"<|start_header_id|>{message["role"]}<|end_header_id|>\n\n{message["content"].strip()}<|eot_id|>"
+    if add_generation_prompt: ret += "<|start_header_id|>assistant<|end_header_id|>\n\n"
+    return ret
+  def generate_until(self, requests: list[Instance]) -> list[str]:
     continuations = []
-    for request in requests:
-      prompt, until = request[0], request[1]['until']
-      output = self.llama.greedy_until(prompt, until, max_length=128, temperature=0.0)
-      continuations.append(output[len(prompt):])
+    for request in tqdm(requests):
+      prompt, args = request.args
+      until = [self.tokenizer.encode(tok) for tok in args.get("until", [])]
+      toks = [self.tokenizer.bos_id] + self.tokenizer.encode(prompt,allow_special=True)
+      prompt_len = len(toks)
+      max_gen_toks = args.get("max_gen_toks") or args.get("max_length") or self.max_length-prompt_len
+      assert self.max_length >= max_gen_toks, "This eval needs a longer context length"
+      start_pos = 0
+      for i in range(max_gen_toks):
+        next_tok = self.model(Tensor([toks[start_pos:]]), start_pos, args.get("temperature", 0.0)).item()
```

Comment:
might be faster to ingest the prompt one token at a time so that the whole thing can be jitted.